tests/Simplify/EqualTest.elm

```diff
@@ -934,6 +934,787 @@ a = ({ a = 1 }).a == ({ a = 1 }).b
+                            |> Review.Test.whenFixed """module A exposing (..)
+a : Bool
+a = (not << List.isEmpty) l
+"""
+                        ]
+        , test "should replace 0 /= List.length l with (not << List.isEmpty)" <|
+            \() ->
+                """module A exposing (..)
+a : Bool
+a = 0 /= List.length l
+"""
+                    |> Review.Test.run ruleWithDefaults
+                    |> Review.Test.expectErrors
+                        [ Review.Test.error
+                            { message = "This can be replaced with a call to (not << List.isEmpty)"
+                            , details =
+                                [ "Whereas List.length takes as long to run as the number of elements in the List,"
+                                , "List.isEmpty runs in constant time."
+                                ]
+                            , under = "0 /= List.length l"
+                            }
+                            |> Review.Test.whenFixed """module A exposing (..)
+a : Bool
+a = (not << List.isEmpty) l
+"""
+                        ]
+        , test "should replace (l |> List.length) /= 0 with (not << List.isEmpty)" <|
+            \() ->
+                """module A exposing (..)
+a : Bool
+a = (l |> List.length) /= 0
+"""
+                    |> Review.Test.run ruleWithDefaults
+                    |> Review.Test.expectErrors
+                        [ Review.Test.error
+                            { message = "This can be replaced with a call to (not << List.isEmpty)"
+                            , details =
+                                [ "Whereas List.length takes as long to run as the number of elements in the List,"
+                                , "List.isEmpty runs in constant time."
+                                ]
+                            , under = "(l |> List.length) /= 0"
+                            }
+                            |> Review.Test.whenFixed """module A exposing (..)
+a : Bool
+a = (l |> (not << List.isEmpty))
+"""
+                        ]
+        , test "should replace 0 /= (l |> List.length) with (not << List.isEmpty)" <|
+            \() ->
+                """module A exposing (..)
+a : Bool
+a = 0 /= (l |> List.length)
+"""
+                    |> Review.Test.run ruleWithDefaults
+                    |> Review.Test.expectErrors
+                        [ Review.Test.error
+                            { message = "This can be replaced with a call to (not << List.isEmpty)"
+                            , details =
+                                [ "Whereas List.length takes as long to run as the number of elements in the List,"
+                                , "List.isEmpty runs in constant time."
+                                ]
+                            , under = "0 /= (l |> List.length)"
+                            }
+                            |> Review.Test.whenFixed """module A exposing (..)
+a : Bool
+a = (l |> (not << List.isEmpty))
```

Comment:
I have to admit I don't like this `not <<` fix. I think it would be much better to wrap in a `not` call, and ideally in a way that follows the existing format:
- `List.length l /= 0` -> `not (List.isEmpty l)`
- `(l |> List.length) /= 0` -> `(l |> List.isEmpty |> not)`
- `(List.length <| l) /= 0` -> `(not <| List.length <| l)`

As far as I can tell, the original position of the operands (`... /= 0` or `0 /= ...`) does not impact the fix.

As for the error message, I think the following would fit:

> This can be replaced with a call to `List.empty` and `not`

The "and `not`" part would only be there when encountering `/=`.

What do you think?

Reply:
is there an existing rule and replacement out there that handles this kind of replacement I could look at?

Reply:
I haven't found one so far. @lue-bird does this ring a bell to you?

Reply:
mapOnWrappedChecks and wrapperMapNChecks have something like that
```elm
case checkInfo.callStyle of
      Pipe LeftToRight ->
          [ Fix.insertAt elementCountCall.parentRange.end
              (" |> " ++ qualifiedToString (qualify Fn.Basics.not checkInfo))
          ]

      Pipe RightToLeft ->
          [ Fix.insertAt elementCountCall.parentRange.start
              (qualifiedToString (qualify Fn.Basics.not checkInfo) ++ " <| ")
          ]

      Application ->
          [ Fix.insertAt elementCountCall.parentRange.end ")"
          , Fix.insertAt elementCountCall.parentRange.start (qualifiedToString (qualify Fn.Basics.not checkInfo) ++ " (")
          ]
```

Reply:
This is really solid, thanks. It's going to take me a little bit of time since I don't have the `callStyle` of the function application here. I may have to add it to `AstHelpers.getSpecificFnCall` in order to have it in this location. The parent `checkInfo` is about the operator application of `==` so that doesn't have it, naturally.